src/unix/core.c

```diff
@@ -414,7 +416,7 @@ int uv_run(uv_loop_t* loop, uv_run_mode mode) {
     uv__run_check(loop);
     uv__run_closing_handles(loop);
 
-    if (mode == UV_RUN_ONCE) {
+    if (mode == UV_RUN_ONCE && did_work == 0) {
```

Comment:
Actually, you could remove all the extra code in each `uv__io_poll()` by caching the `uv__loop_metrics_t.events` value at the top of the while loop then comparing it here. For example, at the top set:
```c
uint64_t events_count = uv__get_loop_metrics(loop)->metrics.events
```

and here just check
```c
events_count == uv__get_loop_metrics(loop)->metrics.events
```

Would still need to check if timers did any work since they're not counted as events.